README.md
# OONI Orchestra

[![Build Status](https://travis-ci.org/ooni/orchestra.svg)](https://travis-ci.org/ooni/orchestra)
[![Release](https://img.shields.io/github/release/ooni/orchestra.svg?style=flat-square)](https://github.com/ooni/orchestra/releases)


The OONI Probe Orchestration System.

This repository contains the various microservices that compose the OONI
Probe Orchestration System.

## Getting started

We use a simple `Makefile`. Please, check it out to see how we invoke `go` in
a farily standard way for building and running tests.

To build all the binaries, run:

```
make orchestra
```

You should now have inside of `./bin` a series of binaries. The ones you care
about are:

```
./bin/ooni-orchestra
./bin/ooni-registry
```

They both take a config file--you can use
`orchestrate/ooni-orchestrate.toml.example` (or the `registry` one) as
examples, respectively.

To aid local testing you should set up a local postgres instance with a user
and a database by running:
```
CREATE USER proteus WITH PASSWORD 'changeme';
CREATE DATABASE proteus;
GRANT ALL PRIVILEGES ON DATABASE proteus to proteus;
```

You should then change the config line in the `database` section for the key
`url` to something like:

```
[database]
url = "postgres://proteus:changeme@127.0.0.1:32768/proteus?sslmode=disable"
```

For both `orchestrate/ooni-orchestrate.toml` and
`registry/ooni-registry.toml`.

You should then be able to start the services by running:

```
./bin/ooni-orchestrate --config orchestrate/ooni-orchestrate.toml start
```

## Components

**registry**

Is responsible for registering probes and keeping tabs on what their related
metadata is.

**frontend**

Is the frontend to:

* Setup schedules

* View active schedules

* View active probes

**orchestrate**

Is responsible for receiving events via the admin interface and triggering
notifications via `gorush`.

Can also be used to view the event history.

## Building and releasing

- Make sure the `GOPATH` environment variable is set. (For reference, the setup
  of sbs is `export GOPATH=$HOME` with repositories in `$HOME/src/`; e.g. this
  repository is `$HOME/src/github.com/ooni/orchestra`).

- Of course, you also need to have golang installed.

- To build for development, run `make orchestra`.

- To create a release, run `make release`.

Checklist before tagging a release:

- Make sure you have updated the changelog

- Make sure you bumped the version number in:

    - `Makefile`

    - `common/orchestra_info.go`

- Make sure all unittests are passing (`make check`)

## Notifications specification

A client needs to register to the proteus-registry service.

The canonical address for it shall be `https://registry.XXX.YYY.ZZZ/`. We
also support the cloudfronted domain as following.

### Registration

Upon first running the app the client needs to registry with the notification
service.

Once a client is registered they can update the various metadata related to the probe by means of an update request (detailed in the following section).

To register a probe a the following HTTPS request is issued:

**Method**: `POST`
**Path**: `/api/v1/register`
**Body**:
```json
{
  "probe_cc": "IT",
  "probe_asn": "AS0",
  "platform": "android",
  "software_name": "ooniprobe-android",
  "software_version": "0.1.1",
  "software_language": "IT",
  "supported_tests": ["tcp_connect", "web_connectivity"],
  "network_type": "wifi",
  "available_bandwidth": "100",
  "token": "TOKEN_ID"
}
```

In particular the `token` field represents the Device Token.

The registration service will return a `client_id` to be used to update the Device Token as well as other metadata.

The response looks like this:

```json
{"client_id": "XXX-YYY-ZZZ-TTT"}
```

### Update metadata

In order do update the metadata you to issue the following request:

**Method**: `PUT`
**Path**: `/api/v1/update/$CLIENT_ID`
**Body**:
```json
{
  "probe_cc": "IT",
  "probe_asn": "AS0",
  "platform": "android",
  "software_name": "ooniprobe-android",
  "software_version": "0.1.1",
  "supported_tests": ["tcp_connect", "web_connectivity"],
  "network_type": "wifi",
  "available_bandwidth": "100",
  "token": "NEW_TOKEN_ID"
}
```

The server will respond with:

```json
{"status": "ok"}
```

in case of an error:

```json
{"error": "ERROR_MESSAGE"}
```

### Notifications

A notification includes in the payload of the silent notification a pingback
URL that the client needs to connect to in order to receive the task that it
need to run to perform the actual measurement.


ChangeLog.md
# orchestra 0.3.2 [2020-01-21]

Changes:

* Go 1.13 support including migrating to Go modules
* Fetch tor targets via authenticated orchestra (#83)

# orchestra 0.3.1 [2019-12-17]

Changes:

* add Dockerfile for the daemons
* add Dockerfile for frontend
* fix bug in list-target query (dfa1d7c527)
* bump frontend dependencies
* distribute psiphon config (#80)

# orchestra 0.3.0 [2019-06-05]

Changes:

* Returns by default all URLs in the test-lists/urls endpoint

# orchestra 0.2.5 [2019-01-22]

Changes:

* Support registering clients with an empty push notification token: https://github.com/ooni/orchestra/issues/41


# orchestra 0.2.4 [2018-12-13]

Changes:

* Validate test URLs when returning them and ignore invalid URLs

Fixes:

* Bug with test URLs endpoint: https://github.com/ooni/orchestra/issues/59
* Bug with test-helpers endpoint: https://github.com/ooni/orchestra/issues/54

# orchestra 0.2.2 [2018-10-22]

Adds:

* test-lists/urls endpoint
* Support for returning full list of URLs with limit=-1

# orchestra 0.2.1 [2018-07-13]

Changes:

* Rename proteus to orchestra
* Switch to dep for dependency management

Adds:

* Pagination support for result listing

# proteus 0.2.0 [2018-01-25]

Fixes:

* Update next.js in proteus-frontend to 3.2.3 to remediate: https://github.com/zeit/next.js/releases/tag/4.2.3 (the fix has been backported to 3.2.3 too)

# proteus 0.2.0-rc.2 [2018-01-16]

Fixes:

* Make unauthenticated endpoints work without auth header

# proteus 0.2.0-rc.1 [2017-12-14]

Changes:

* Rename proteus-events to proteus-orchestrate
* Remove the deprecated proteus-notify

Adds:

* New orchestrate endpoints for /test-helpers, /collectors and /urls

# proteus 0.1.0-rc.1 [2017-09-29]

Changes:

* Improvements to testing and code style

* Add unittests that speak to a database

# proteus 0.1.0-beta.9 [2017-08-06]

* fix(registry): Language may be null
* fix(database): inconsistent migrations

# proteus 0.1.0-beta.8 [2017-08-02]

* refactor(Makefile): simplify the proteus targets dependencies
* feature(README.md): build and release instruction
* feature(events): fwd `click_action` for Android
* fix(registry): adjust syntax of `add_language_column`
* fix(Makefile): don't assume a tool has bindata
* regen(bindata): run `make bindata`
* fix(Makefile): always update embedded binary data

# proteus 0.1.0-beta.7 [2017-07-30]

* fix(proteus-events): do not propagate topic on Android

# proteus 0.1.0-beta.6 [2017-07-24]

* Improvements to the scheduling UI

* Add minimal progress bar

# proteus 0.1.0-beta.5 [2017-07-19]

* Fix bug in migration script

* Re-enable measurement job scheduling from UI

# proteus 0.1.0-beta.4 [2017-07-17]

* Fix bug in deletion of jobs

# proteus 0.1.0-beta.3 [2017-06-23]

* Add support for alert notifications

* Temporarily disable measurement job scheduling from UI

* Use gorush instead of proteus-notify

# proteus 0.1.0-beta.2 [2017-05-28]

* Fix the schema of proteus-events

# proteus 0.1.0-beta.1 [2017-05-26]

* Add support for Admins to manage jobs via the web UI

# proteus 0.1.0-dev [2017-05-13]

Development release of proteus.

Includes:

* Ability for probes to register with the orchestration registry
* Ability for probes to update the metadata about their probe
* Support for sending push notifications via Apple Push Notifcation service and Firebase Cloud Messagging
* Support for administrators to schedule measurements via a web interface
* Ability for probes to receive the tasks they have been notified about and mark them as accepted, rejected or done
* Multiple architecture build system



cmd/start.go
package cmd

import (
	"fmt"
	"time"
	"errors"
	"net/http"
	"database/sql"

	"github.com/satori/go.uuid"
	"github.com/lib/pq"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
	"gopkg.in/gin-gonic/gin.v1"
	_ "github.com/facebookgo/grace/gracehttp"
)

// startCmd represents the start command
var startCmd = &cobra.Command{
	Use:   "start",
	Short: "Start the server",
	Long: `A longer description that spans multiple lines and likely contains examples
and usage of using your command. For example:

Cobra is a CLI library for Go that empowers applications.
This application is a tool to generate the needed files
to quickly create a Cobra application.`,
	Run: func(cmd *cobra.Command, args []string) {
		start()
	},
}

func initDatabase() (*sql.DB, error) {
	db, err := sql.Open("postgres", viper.GetString("database-url"))
	if err != nil {
		ctx.Error("failed to open database")
		return nil, err
	}
	return db, err
}

type RegisterReq struct {
	ProbeCC string `json:"probe_cc" binding:"required"`
	ProbeASN string `json:"probe_asn" binding:"required"`
	Platform string `json:"platform" binding:"required"`

	SoftwareName string `json:"software_name" binding:"required"`
	SoftwareVersion string `json:"software_version" binding:"required"`
	SupportedTests []string `json:"supported_tests"`

	NetworkType string `json:"network_type"`
	AvailableBandwidth string `json:"available_bandwidth"`
	
	DeviceToken string `json:"device_token"`

	ProbeFamily string `json:"probe_family"`
	ProbeID string `json:"probe_id"`
}

type UpdateReq struct {
	MetadataID string `json:"metadata_id" binding:"required"`
	ProbeCC string `json:"probe_cc"`
	ProbeASN string `json:"probe_asn"`
	Platform string `json:"platform"`

	SoftwareName string `json:"software_name"`
	SoftwareVersion string `json:"software_version"`
	SupportedTests []string `json:"supported_tests"`

	NetworkType string `json:"network_type"`
	AvailableBandwidth string `json:"available_bandwidth"`
	
	DeviceToken string `json:"device_token"`

	ProbeFamily string `json:"probe_family"`
	ProbeID string `json:"probe_id"`
}

type HeartbeatReq struct {
	MetadataID string `json:"metadata_id"`
	LastHeartbeat string `json:""`
}

func Update(db *sql.DB, req UpdateReq) (error) {
	query := fmt.Sprintf(`INSERT INTO %s (
		id, update_time,
		metadata_id,
		probe_cc, probe_asn,
		platform, software_name,
		software_version, supported_tests,
		network_type, available_bandwidth,
		device_token, probe_family,
		probe_id, update_type
	) VALUES (
		$1, $2,
		$3, $4,
		$5, $6,
		$7, $8,
		$9, $10,
		$11, $12,
		$13, $14, $15)`,
		pq.QuoteIdentifier(viper.GetString("probe-updates-table")))

	id := uuid.NewV4().String()
	_, err := db.Exec(query,
						id, time.Now().UTC(),
						req.MetadataID,
						req.ProbeCC, req.ProbeASN,
						req.Platform, req.SoftwareName,
						req.SoftwareVersion, pq.Array(req.SupportedTests),
						req.NetworkType, req.AvailableBandwidth,
						req.DeviceToken, req.ProbeFamily,
						req.ProbeID, "update")
	if (err != nil) {
		ctx.WithError(err).Error("failed to add data to update table")
		return errors.New("error in adding data to update probes")
	}
	return nil
}

func Register(db *sql.DB, req RegisterReq) (string, error) {
	if (req.Platform == "ios" && req.DeviceToken == "") {
		return "", errors.New("missing device token")
	}

	// XXX Wrap this all in a transaction
	// figure out what this means: https://github.com/lib/pq/issues/81
	query := fmt.Sprintf(`INSERT INTO %s (
		id, creation_time,
		last_updated,
		probe_cc, probe_asn,
		platform, software_name,
		software_version, supported_tests,
		network_type, available_bandwidth,
		device_token, probe_family,
		probe_id
	) VALUES (
		$1, $2,
		$3, $4,
		$5, $6,
		$7, $8,
		$9, $10,
		$11, $12,
		$13, $14)`,
		pq.QuoteIdentifier(viper.GetString("active-probes-table")))

	metadataID := uuid.NewV4().String()
	_, err := db.Exec(query,
						metadataID, time.Now().UTC(),
						time.Now().UTC(),
						req.ProbeCC, req.ProbeASN,
						req.Platform, req.SoftwareName,
						req.SoftwareVersion, pq.Array(req.SupportedTests),
						req.NetworkType, req.AvailableBandwidth,
						req.DeviceToken, req.ProbeFamily,
						req.ProbeID)
	if (err != nil) {
		ctx.WithError(err).Error("failed to add data to active probe table")
		return "", errors.New("error in adding data to active probes")
	}

	query = fmt.Sprintf(`INSERT INTO %s (
		id, update_time,
		metadata_id,
		probe_cc, probe_asn,
		platform, software_name,
		software_version, supported_tests,
		network_type, available_bandwidth,
		device_token, probe_family,
		probe_id, update_type
	) VALUES (
		$1, $2,
		$3, $4,
		$5, $6,
		$7, $8,
		$9, $10,
		$11, $12,
		$13, $14, $15)`,
		pq.QuoteIdentifier(viper.GetString("probe-updates-table")))

	updateID := uuid.NewV4().String()
	_, err = db.Exec(query,
						updateID, time.Now().UTC(),
						metadataID,
						req.ProbeCC, req.ProbeASN,
						req.Platform, req.SoftwareName,
						req.SoftwareVersion, pq.Array(req.SupportedTests),
						req.NetworkType, req.AvailableBandwidth,
						req.DeviceToken, req.ProbeFamily,
						req.ProbeID, "register")
	if (err != nil) {
		ctx.WithError(err).Error("failed to add data to update table")
		return "", errors.New("error in adding data to update probes")
	}

	return metadataID, nil
}

func start() {
	db, err := initDatabase()

	if (err != nil) {
		ctx.WithError(err).Error("failed to connect to DB")
		return
	}
	defer db.Close()

	router := gin.Default()
	router.POST("/api/v1/register", func(c *gin.Context) {
		var registerReq RegisterReq
		err := c.BindJSON(&registerReq)
		if (err != nil) {
			ctx.WithError(err).Error("invalid request")
			c.JSON(http.StatusBadRequest,
					gin.H{"error": "invalid request"})
			return
		}

		metadataID, err := Register(db, registerReq)
		if (err != nil) {
			c.JSON(http.StatusBadRequest,
					gin.H{"error": err.Error()})
			return
		}

		c.JSON(http.StatusOK, gin.H{"metadata_id": metadataID})
		return
	})
	router.POST("/api/v1/update", func(c *gin.Context) {
		var updateReq UpdateReq
		err := c.BindJSON(&updateReq)
		if (err != nil) {
			ctx.WithError(err).Error("invalid request")
			c.JSON(http.StatusBadRequest,
					gin.H{"error": "invalid request"})
			return
		}
		err = Update(db, updateReq)
		if (err != nil) {
			ctx.WithError(err).Error("failed to update")
			c.JSON(http.StatusBadRequest,
					gin.H{"error": err.Error()})
			return
		}
	})

	router.POST("/api/v1/heartbeat", func(c *gin.Context) {
		//var heartbeatReq HeartbeatReq
	})
	
	Addr := fmt.Sprintf("%s:%d", viper.GetString("server-address"),
								viper.GetInt("server-port"))
	ctx.Infof("starting on %s", Addr)
	s := &http.Server{
		Addr: Addr,
		Handler: router,
	}
	//gracehttp.Serve(s)
	s.ListenAndServe()
}

func init() {
	RootCmd.AddCommand(startCmd)

	// Here you will define your flags and configuration settings.

	// Cobra supports Persistent Flags which will work for this command
	// and all subcommands, e.g.:
	startCmd.PersistentFlags().IntP("server-port", "", 8080, "Which port we should bind to")
	startCmd.PersistentFlags().StringP("server-address", "", "127.0.0.1", "Which interface we should listen on")
	viper.BindPFlag("server-port", startCmd.PersistentFlags().Lookup("server-port"))
	viper.BindPFlag("server-address", startCmd.PersistentFlags().Lookup("server-address"))
}
